// @crowdin/crowdin-api-client/out/issues/index.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.Issues = void 0;
const core_1 = require("../core");
/**
 * @deprecated
 * @ignore
 */
class Issues extends core_1.CrowdinApi {
    listReportedIssues(projectId, options, deprecatedOffset, deprecatedType, deprecatedStatus) {
        if ((0, core_1.isOptionalNumber)(options, '1' in arguments)) {
            options = {
                limit: options,
                offset: deprecatedOffset,
                type: deprecatedType,
                status: deprecatedStatus,
            };
        }
        let url = `${this.url}/projects/${projectId}/issues`;
        url = this.addQueryParam(url, 'type', options.type);
        url = this.addQueryParam(url, 'status', options.status);
        return this.getList(url, options.limit, deprecatedOffset);
    }
    /**
     * @param projectId project identifier
     * @param issueId issue identifier
     * @param request request body
     * @see https://support.crowdin.com/api/v2/#operation/api.projects.issues.patch
     */
    editIssue(projectId, issueId, request) {
        const url = `${this.url}/projects/${projectId}/issues/${issueId}`;
        return this.patch(url, request, this.defaultConfig());
    }
}
exports.Issues = Issues;
